src/nvim/tui/tui.c

```diff
@@ -406,52 +471,65 @@ static void tui_main(UIBridgeData *bridge, UI *ui)
 #if TERMKEY_VERSION_MAJOR > 0 || TERMKEY_VERSION_MINOR > 18
   data->input.tk_ti_hook_fn = tui_tk_ti_getstr;
 #endif
-  tinput_init(&data->input, &tui_loop);
+  tinput_init(&data->input, &main_loop);
   tui_terminal_start(ui);
+  loop_schedule(&main_loop, event_create(show_termcap_event, 1, data->ut));
 
-  // Allow main thread to continue, we are ready to handle UI callbacks.
-  CONTINUE(bridge);
+  // // "Active" loop: first ~100 ms of startup.
+  // for (size_t ms = 0; ms < 100 && !tui_is_stopped(ui);) {
+  //   ms += (loop_poll_events(&main_loop, 20) ? 20 : 1);
+  // }
+}
 
-  loop_schedule_deferred(&main_loop,
-                         event_create(show_termcap_event, 1, data->ut));
+void tui_execute(void) {
+  UI *ui = get_ui_by_index(1);
+  LOOP_PROCESS_EVENTS(&main_loop, main_loop.events, -1);
+  tui_io_driven_loop(ui);
+  tui_exit_safe(ui);
+  // getout((int)server_process_exit_status);
+  getout(0);
+}
 
-  // "Active" loop: first ~100 ms of startup.
-  for (size_t ms = 0; ms < 100 && !tui_is_stopped(ui);) {
-    ms += (loop_poll_events(&tui_loop, 20) ? 20 : 1);
-  }
+// Doesn't return until the TUI is closed (by call of tui_stop())
+void tui_io_driven_loop(UI *ui){
```

Comment:
could be `static`

Reply:
Except `tui_execute` almost all functions will be made `static` in tui.c